sql/src/main/java/io/crate/metadata/cluster/AlterTableClusterStateExecutor.java

```diff
@@ -0,0 +1,437 @@
+    private final MetaDataCreateIndexService metaDataCreateIndexService;
+
+    public AlterTableClusterStateExecutor(MetaDataMappingService metaDataMappingService,
+                                          IndicesService indicesService,
+                                          AllocationService allocationService,
+                                          IndexScopedSettings indexScopedSettings,
+                                          IndexNameExpressionResolver indexNameExpressionResolver,
+                                          MetaDataCreateIndexService metaDataCreateIndexService) {
+        this.metaDataMappingService = metaDataMappingService;
+        this.indicesService = indicesService;
+        this.indexScopedSettings = indexScopedSettings;
+        this.allocationService = allocationService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.metaDataCreateIndexService = metaDataCreateIndexService;
+    }
+
+    @Override
+    protected ClusterState execute(ClusterState currentState, AlterTableRequest request) throws Exception {
+        Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(currentState,
+                                                                              INDICES_OPTIONS,
+                                                                              request.tableIdent().indexNameOrAlias());
+        if (request.isPartitioned()) {
+            if (request.partitionIndexName() != null) {
+                concreteIndices = indexNameExpressionResolver.concreteIndices(currentState,
+                                                                              INDICES_OPTIONS,
+                                                                              request.partitionIndexName());
+                currentState = updateMapping(currentState, request, concreteIndices);
+                currentState = updateSettings(currentState, request.settings(), concreteIndices);
+            } else {
+                // template gets all changes unfiltered
+                currentState = updateTemplate(currentState, request);
+
+                if (!request.excludePartitions()) {
+
+                    // These settings only apply for already existing partitions
+                    List<String> supportedSettings = TableParameters.PARTITIONED_TABLE_PARAMETER_INFO_FOR_TEMPLATE_UPDATE
+                        .supportedSettings()
+                        .values()
+                        .stream()
+                        .map(Setting::getKey)
+                        .collect(Collectors.toList());
+
+                    // auto_expand_replicas must be explicitly added as it is hidden under NumberOfReplicasSetting
+                    supportedSettings.add(IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS);
+                    TableParameter parameterWithFilteredSettings = new TableParameter(request.settings(), supportedSettings);
```

Comment:
This looks like an abuse of the `TableParameter` class to filter settings like this. Can we change this? extract the filtering into a separate method maybe.